src/main/java/ch/uzh/ifi/hase/soprafs23/controller/GroupController.java

```diff
@@ -223,5 +223,50 @@ public List<UserGetDTO> getGroupMembersById(@PathVariable Long groupId, HttpServ
 
         return userGetDTOs;
     }
-    
+    @DeleteMapping("/groups/{groupId}")
+    @ResponseStatus(HttpStatus.NO_CONTENT) // 204
+    public void deleteGroup(@PathVariable Long groupId, HttpServletRequest request) {
+        // Check if the group exists
+        Group group = groupService.getGroupById(groupId); // 404 - group not found
+
+        // Check the validity of the token
+        Long tokenId = userService.getUseridByToken(request.getHeader("X-Token"));
+        if (tokenId.equals(0L)) {
+            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Not authorized"); // 401 - not authorized
+        }
+
+        // Check if the user is the host of the group
+        if (!tokenId.equals(group.getHostId())) {
+            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Not authorized"); // 401 - not authorized
+        }
+
+        // Delete the group
+        groupService.deleteGroup(groupId);
+    }
+
+    @PutMapping("/groups/{groupId}/leave")
+    @ResponseStatus(HttpStatus.NO_CONTENT) // 204
+    public void leaveGroup(@PathVariable Long groupId, HttpServletRequest request) {
+        // Check if the group exists
+        Group group = groupService.getGroupById(groupId); // 404 - group not found
+
+        // Check the validity of the token
+        Long guestId = userService.getUseridByToken(request.getHeader("X-Token"));
+        if (guestId.equals(0L)) {
+            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Not authorized"); // 401 - not authorized
+        }
+
+        // Check if the user is a member of the group
+        List<Long> memberIds = groupService.getAllMemberIdsOfGroup(group);
+        if (!memberIds.contains(guestId)) {
+            throw new ResponseStatusException(HttpStatus.CONFLICT, "Guest was not in the group"); // 409 - conflict
+        }
+
```

Comment:
PROBLEM! check if the tokenId corresponds to the hostId of the group! atm the host can leave the group and it's very weird because in the group they are still the host but in the users they are in no group and all the guests stay in the group ^^' 
I'd say throw a 409 in that case with something like "you are the host. if you want to leave please delete the group" or something like that

Reply:
like the error suggestion!